consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/FutureMessageBuffer.java

```diff
@@ -0,0 +1,101 @@
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft.statemachine;
+
+import tech.pegasys.pantheon.ethereum.p2p.api.Message;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Buffer which holds future IBFT messages.
+ *
+ * <p>This buffer only allows messages to be added which have a chain height greater than current
+ * height and up to chain futureMessagesMaxDistance from the current chain height.
+ *
+ * <p>When the total number of messages is greater futureMessagesLimit then messages are evicted.
+ *
+ * <p>If there is more than one height in the buffer then all messages for the highest chain height
+ * are removed. Otherwise if there is only one height the oldest inserted message is removed.
+ */
+public class FutureMessageBuffer {
+  private final NavigableMap<Long, List<Message>> buffer;
+  private final long futureMessagesMaxDistance;
+  private final long futureMessagesLimit;
+  private final Supplier<Long> currentChainHeight;
+  private long totalMessages;
+
+  public FutureMessageBuffer(
+      final long futureMessagesMaxDistance,
+      final long futureMessagesLimit,
+      final Supplier<Long> currentChainHeight) {
+    this(futureMessagesMaxDistance, futureMessagesLimit, currentChainHeight, new TreeMap<>());
+  }
+
+  @VisibleForTesting
+  FutureMessageBuffer(
+      final long futureMessagesMaxDistance,
+      final long futureMessagesLimit,
+      final Supplier<Long> currentChainHeight,
+      final NavigableMap<Long, List<Message>> buffer) {
+    this.futureMessagesMaxDistance = futureMessagesMaxDistance;
+    this.futureMessagesLimit = futureMessagesLimit;
+    this.currentChainHeight = currentChainHeight;
+    this.buffer = buffer;
+  }
+
+  public void addMessage(final long msgChainHeight, final Message rawMsg) {
+    final long currentHeight = currentChainHeight.get();
+    if (!validMessageHeight(msgChainHeight, currentHeight)) {
+      return;
+    }
+
+    if (totalMessages >= futureMessagesLimit) {
+      evictMessages();
+    }
+
+    if (!buffer.containsKey(msgChainHeight)) {
+      buffer.put(msgChainHeight, new ArrayList<>());
+    }
+    buffer.get(msgChainHeight).add(rawMsg);
+    totalMessages += 1;
+  }
+
+  private boolean validMessageHeight(final long msgChainHeight, final long currentHeight) {
+    final boolean isFutureMsg = msgChainHeight > currentHeight;
+    final boolean withinMaxChainHeight =
+        msgChainHeight <= currentHeight + futureMessagesMaxDistance;
+    return isFutureMsg && withinMaxChainHeight;
+  }
+
+  private void evictMessages() {
+    if (buffer.size() > 1) {
+      buffer.remove(buffer.lastKey());
+    } else {
+      List<Message> messages = buffer.firstEntry().getValue();
```

Comment:
Is it possible that buffer size is 0? I guess not.

Reply:
I am making an assumption that it isn't 0 at this point but that only holds if futureMessagesLimit isn't 0. Be better if the evict messages handled that case, so i'll update evictMessages to handle that properly.

Reply:
done. added handling of zero message limit.